### PR TITLE
Recipe App

### PR DESCRIPTION
**Pull request by Miliyonayalew**
# In this Final Rails Group Capstone Project we @IngDavidTM  and @Miliyonayalew Built a Recipe App that has the following requirements
- Used Ruby gems as a software packages system.
- Installed Ruby on Rails framework.
- Understand Rails RESTful design and router.
- Used controllers to handle requests and render empty views.
- Understood Rails naming conventions.
- Used params from browser requests in a safe way.
- Used preprocessed HTML file with embedded Ruby code.
- Used layouts and templates for shared content.
- Used database migration files to maintain database schema.
- Used validations for models.
- Secured app from n+1 problems.
- Understand what ORM is.
- Wrote SQL queries with ActiveRecord.
- We Set up associations between models.
- Build a web app that requires the user to log in.
- Describe the difference between authorization and authentication.
- Used Devise gem for authentication.
- Understood how sessions and cookies can support authentication.
- Limited access to web app resources based on authorization rules.

# Requirements that we did
A login page.
A registration page.
A food list (with all CRUD implementation, except for 'update').
A recipe list (with all CRUD implementation, except for 'update').
A general shopping list view (all missing food for all your recipes and total price).
A list of all public recipes from other users with their names and total prices.

### CODE REVIEW

File: app/views/recipes/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      <%=link_to "Add ingredient", new_recipe_recipe_food_path(@recipe), class: "button" %>
+    <%end%>
+  </div>
+  <table class="table">
+      <thead>
+        <tr>
+          <th>Food</th>
+          <th>Quantity</th>
+          <th>Value</th>
+          <th>Actions</th>
```
Comment (t-yanick) [inline app/views/recipes/show.html.erb:26]: - [ ] Great job working on this app. It is quite awesome. As per the wire frame when a user clicks on a recipe item, it goes to the recipe details page Here under actions after adding ingredients, there should be Modify and Remove but as per the screenshot below, your app only has Remove and as such a user cannot Modify an addes ingredient Kindly add this action
image

File: spec/models/food_spec.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  context 'Testing validations' do
+    it 'Validation should be successful' do
+      expect(@food).to be_valid
+    end
+
+    it 'Food should have a name' do
+      expect(@food.name).to be_present
+    end
+
```
Comment (t-yanick) [inline spec/models/food_spec.rb:20]: - [ ] Good job adding both unit and integration tests. Unfortunately, some of your test cases are failing as seen in the screenshot below. Kindly fix these failing tests.
image

**Review by t-yanick — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED

Hi @Miliyonayalew and @IngDavidTM , 

### Good job so far!

### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Great use of partials for views
:heavy_check_mark: N + 1 problem fixed
:heavy_check_mark: Good respect of wireframes
:heavy_check_mark: Devise used for login and registration
:heavy_check_mark: Rspec and Capybara used for testing
:heavy_check_mark: Excellent documentation
:heavy_check_mark: Necessary logic put in the controllers

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/food_spec.rb (hunk 2 above)
Comment (IngDavidTM) [reply spec/models/food_spec.rb:20]: Hi @t-yanick, those tests are passing on both, my computer and my partner's computer. We have tried everything and it is working. I will do a video to validate that the testing is correct.

Comment (IngDavidTM) [reply spec/models/food_spec.rb:20]: https://user-images.githubusercontent.com/105117832/217973918-9901cd88-f985-4df7-83a5-9a834cf8ef60.mp4

Comment (t-yanick) [reply spec/models/food_spec.rb:20]: - [ ] Hi Team, great job. You can request for a re-review and the next reviewer will handle it.

**Review by 1GORDON — APPROVED** [review]
# Status: APPROVED 🟢 

Hi @Miliyonayalew & @IngDavidTM ✋

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! :tada: :clap: :clap: :100:

## Highlights!
- Followed Gitflow ✔️
- Good commit messages ✔️
- Linters are passing ✔️
- Descriptive Pull request ✔️
- Professional readme ✔️

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread and tag me `@1GORDON` if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/food_spec.rb (hunk 2 above)
Comment (Miliyonayalew) [reply spec/models/food_spec.rb:20]: Hey @t-yanick, Thanks for the review and feedback

**Comment by Miliyonayalew** [comment]
Hey @1GORDON, Thanks for the review

**Pull request merged**